File: FlightData.scala
import scala.io.BufferedSource
import java.io.PrintWriter
import java.io.File

case class Flight(flightID : Int = 0, 
from : String = "", to : String = "", date : Date = Date())

case class Date(year : Int = 0, month : Int = 0, day : Int = 0) {

    def toDays() : Int = {

        this.year * 1000 + this.month * 12 * this.day;
    }
}

case class Passenger(passengerID : Int = 0, 
firstName : String = "", lastName : String = "")

case class PassengerFlight(passengerID : Int = 0, 
flightID : Int = 0)

case class PassengerPair(passengerIDA : Int = 0, 
passengerIDB : Int = 0)

object FlightData {

    val passengerFilename = "passengers.csv";
    val flightDataFilename = "flightData.csv";

    val question1Filename = "scalaQuestion1.csv";
    val question2Filename = "scalaQuestion2.csv";
    val question3Filename = "scalaQuestion3.csv";
    val question4Filename = "scalaQuestion4.csv";
    val question5Filename = "scalaQuestion5.csv";

    def main(args : Array[String]) = {

        println(s"Reading $passengerFilename");
        val passengers = getPassengers(passengerFilename);

        println(s"Reading $flightDataFilename");
        val flights = getFlights(flightDataFilename);
        val passengerFlights = getPassengerFlights(flightDataFilename);

        val passengerIDFlights 
        = getPassengerIDFlights(flights, passengerFlights);

        println(s"Solving Question 1");
        val totalFlightsForEachMonth = getTotalFlightsForEachMonth(
        passengers, flights, question1Filename);

        outputTotalFlightsForEachMonth(totalFlightsForEachMonth, 
        question1Filename);

        println(s"Solving Question 2");
        val mostFrequentFlyers = getMostFrequentFlyers(
        passengers, flights, passengerIDFlights, 100);

        outputMostFrequentFlyers(mostFrequentFlyers, 
        question2Filename);

        println(s"Solving Question 3");
        val passengerRuns = getPassengerRuns(passengers, flights, 
        passengerIDFlights, "uk");

        outputPassengerRuns(passengerRuns, question3Filename);

        println(s"Solving Question 4");
        val passengerPairs = getPassengerPairs(passengers);
        val passengerPairFlights = getPassengerPairFlights(
        passengerPairs, passengerIDFlights);
        val passengerPairNumberOfFlights 
        = getPassengerPairNumberOfFlights(passengerPairFlights, 3);

        outputPassengerPairNumberOfFlights(passengerPairNumberOfFlights, 
        question4Filename);
    }

    def getPassengers(filename : String) : Map[Int, Passenger] = {

        // Returns a map of (passengerID, passenger) pairs 
        // from the given input filename 

        val inputFile : BufferedSource = io.Source.fromFile(filename);
        
        inputFile.getLines.drop(1).map({
            case s"$passengerID,$firstName,$lastName" 
                => (passengerID.toInt 
                -> Passenger(passengerID.toInt, firstName, lastName))
        }).toMap;
    }

    def getFlights(filename : String) : Map[Int, Flight] = {

        // Returns a map of (flightID, flight) pairs 
        // from the given input filename 

        val inputFile : BufferedSource = io.Source.fromFile(filename);
        
        inputFile.getLines.drop(1).map({
            case s"$passengerID,$flightID,$from,$to,$year-$month-$day" 
                => (flightID.toInt 
                -> Flight(flightID.toInt, from, to, 
                Date(year.toInt, month.toInt, day.toInt)))
        }).toMap;
    }

    def getPassengerFlights(filename : String) : Set[PassengerFlight] = {

        // Returns a set of passenger flights 
        // representing each passenger travel 
        // from the given input filename 

        val inputFile : BufferedSource = io.Source.fromFile(filename);
        
        inputFile.getLines.drop(1).map({
            case s"$passengerID,$flightID,$from,$to,$date" 
                => PassengerFlight(passengerID.toInt, flightID.toInt)
        }).toSet;
    }

    def getPassengerIDFlights(
    flights : Map[Int, Flight], 
    passengerFlights : Set[PassengerFlight]) : Map[Int, List[Int]] = {

        // Returns a map of (passengerID, list of flightIDs) pairs 
        // from the given set of passenger flights 
        // The list of flights is sorted by date in ascending order 

        passengerFlights.groupBy(passengerFlight 
        => passengerFlight.passengerID)
        .mapValues(passengerFlights => passengerFlights.toList.map(
            passengerFlight => passengerFlight.flightID
        ).sortBy(flightID => flights.getOrElse(flightID, Flight())
        .date.toDays())
        ).toMap
    }

    def getPassengerPairs(
    passengers : Map[Int, Passenger]) : Set[PassengerPair] = {

        // Returns a set of passenger pairs 
        // given a set of passengerIDs 

        // Ensures passengerIDA < passengerIDB to avoid 
        // double counting pairs, as well as avoiding 
        // pairs composed of of the same passenger 
        passengers.keySet.flatMap(passengerIDA 
        => passengers.keySet.map(passengerIDB 
        => PassengerPair(passengerIDA, passengerIDB)))
        .filter(passengerPair 
        => passengerPair.passengerIDA < passengerPair.passengerIDB)
    }

    def getPassengerPairFlights(
    passengerPairs : Set[PassengerPair], 
    passengerIDFlights : Map[Int, List[Int]]) 
    : Map[PassengerPair, Set[Int]] = {

        // Returns a map of (passengerPair, 
        // set of flights shared by the passengerPair) 
        // given a set of passenger pairs 
        // and the flights travelled on by each passenger 

        // Create sets of flights rather than lists for more 
        // time efficient O(1) contains methods 
        val passengerIDFlightSet = passengerIDFlights
        .mapValues(flights => flights.toSet)

        passengerPairs.map(passengerPair 
        => (passengerPair 
            -> passengerIDFlightSet.getOrElse(passengerPair
            .passengerIDA, Set())
            .union(passengerIDFlightSet.getOrElse(passengerPair
            .passengerIDB, Set()))
        )).toMap
    }

    def getTotalFlightsForEachMonth(
    passengers : Map[Int, Passenger], 
    flights : Map[Int, Flight], 
    filename : String) : List[(Int, Int)] = {

        flights.values.groupBy(flight => flight.date.month)
        .mapValues(group => group.size).toList
        .sortBy((month, numberOfFlights) => month);
    }

    def outputTotalFlightsForEachMonth(
    totalFlightsForEachMonth : List[(Int, Int)], 
    filename : String) = {

        // Outputs the total flights for each month 
        // to the given output filename 

        val writer = new PrintWriter(new File(filename));

        writer.println("month,numberOfFlights");
        totalFlightsForEachMonth.map((month, numberOfFlights) 
        => s"$month,$numberOfFlights").foreach({
            line => writer.println(line);
        })

        writer.close();
    }

    def getMostFrequentFlyers(
    passengers : Map[Int, Passenger], 
    flights : Map[Int, Flight], 
    passengerIDFlights : Map[Int, List[Int]], 
    numberOfFlyers : Int) : List[(Passenger, Int)] = {

        // Returns a list of the numberOfFlyers 
        // most frequent flyers 
        // sorted by number of flights in descending order 

        passengers.values.map(passenger => (passenger, 
        passengerIDFlights.mapValues(flights => flights.size)
        .getOrElse(passenger.passengerID, 0)))
        .toList.sortBy((passenger, numberOfFlights) => -numberOfFlights)
        .take(numberOfFlyers);
    }

    def outputMostFrequentFlyers(
    mostFrequentFlyers : List[(Passenger, Int)], filename : String) = {

        // Outputs the k most frequent flyers 
        // sorted by number of flights in descending order 
        // to the given output filename 

        val writer = new PrintWriter(new File(filename));

        writer.println("passengerId,numberOfFlights,firstName,lastName");
        mostFrequentFlyers.map((passenger, numberOfFlights) 
        => (passenger.passengerID, numberOfFlights, 
        passenger.firstName, passenger.lastName))
        .map((passengerID, numberOfFlights, firstName, lastName) 
        => s"$passengerID,$numberOfFlights,$firstName,$lastName")
        .foreach({
            line => writer.println(line);
        })

        writer.close();
    }

    def getPassengerRuns(
    passengers : Map[Int, Passenger], 
    flights : Map[Int, Flight], 
    passengerIDFlights : Map[Int, List[Int]], 
    runEnd : String) : List[(Int, Int)] = {

        // Returns the longest runs for each passenger 
        // sorted by longest run in descending order 
        // given a location runEnd 

        // Ensure the start location is included with the first 
        // from location, followed by all to locations 
        passengerIDFlights.mapValues(
        flightIDs 
        => flightIDs.map(
            flightID => flights.getOrElse(flightID, Flight()).from)
            .take(1).concat(
            flightIDs.map(
            flightID => flights.getOrElse(flightID, Flight()).to))
        .reduce((a, b) => s" $a $b ").split(s" $runEnd ")
        .map(run => run.split("\\s+")
        .filter(location => location.size > 0).size)
        .reduce((a, b) => Math.max(a, b)))
        .toList.sortBy((passengerID, longestRun) => -longestRun)
    }

    def outputPassengerRuns(
    passengerRuns : List[(Int, Int)], filename : String) = {

        // Outputs the longest runs for each passenger 
        // sorted by longest run in descending order 
        // to the given output filename 

        val writer = new PrintWriter(new File(filename));

        writer.println("passengerId,longestRun");
        passengerRuns.map((passengerID, longestRun) 
        => s"$passengerID,$longestRun")
        .foreach({
            line => writer.println(line);
        })

        writer.close();
    }

    def getPassengerPairNumberOfFlights(
    passengerPairFlights : Map[PassengerPair, Set[Int]], 
    minSharedFlights : Int) : List[(Int, Int, Int)] = {

        // Returns the number of flights shared by 
        // each passenger pair only if > and not = minSharedFlights 
        // sorted by number of flights shared in descending order 

        passengerPairFlights.toList
        .map((passengerPair, flights) 
        => (passengerPair.passengerIDA, 
        passengerPair.passengerIDB, flights.size))
        .filter((a, b, numberOfFlights) 
        => (numberOfFlights > minSharedFlights))
        .sortBy((a, b, numberOfFlights) => numberOfFlights)
    }

    def outputPassengerPairNumberOfFlights(
    passengerPairNumberOfFlights : List[(Int, Int, Int)], 
    filename : String) = {

        // Outputs the number of flights shared by 
        // each passenger pair 
        // sorted by number of flights shared in descending order 
        // to the given output filename 

        val writer = new PrintWriter(new File(filename));

        writer.println("passenger1Id,passenger2Id,numberOfFlightsTogether");
        passengerPairNumberOfFlights.map(
        (passenger1ID, passenger2Id, numberOfFlights) 
        => s"$passenger1ID,$passenger2Id,$numberOfFlights")
        .foreach({
            line => writer.println(line);
        })

        writer.close();
    }
}

File: FlightData.java

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Scanner;
import java.util.Set;

class Flight implements Comparable<Flight> {

    public int flightID;
    public String from;
    public String to;
    public Date date;

    public Set<Passenger> passengers = new HashSet<>();

    public Flight(int flightID, 
    String from, String to, Date date) {

        this.flightID = flightID;
        this.from = from;
        this.to = to;
        this.date = date;
    }

    public int compareTo(Flight other) {

        // Sort flights by date ascending 
        return this.date.compareTo(other.date);
    }

    public boolean equals(Flight other) {

        return (this.flightID == other.flightID);
    }
}

class Date implements Comparable<Date> {

    public int year;
    public int month;
    public int day;

    public Date(int year, int month, int day) {

        this.year = year;
        this.month = month;
        this.day = day;
    }

    public int compareTo(Date other) {

        // Sort by year then month then day ascending 
        if (this.year != other.year) {

            return (this.year - other.year);
        }

        if (this.month != other.month) {

            return (this.month - other.month);
        }

        return (this.day - other.day);
    }
}

class Passenger implements Comparable<Passenger> {

    public int passengerID;
    public String firstName;
    public String lastName;

    public Set<Flight> flights = new HashSet<>();

    public Passenger(int passengerID, 
    String firstName, String lastName) {

        this.passengerID = passengerID;
        this.firstName = firstName;
        this.lastName = lastName;
    }

    public int compareTo(Passenger other) {

        // Sort passengers by number of flights ascending 
        return (this.flights.size() - other.flights.size());
    }
}

class PassengerRun implements Comparable<PassengerRun> {

    public int passengerID;
    public int longestRun;

    public PassengerRun(int passengerID, int longestRun) {

        this.passengerID = passengerID;
        this.longestRun = longestRun;
    }

    public int compareTo(PassengerRun other) {

        // Sorts passenger runs by longest run ascending 
        return (this.longestRun - other.longestRun);
    }
}

class PassengerPair implements Comparable<PassengerPair> {

    public int passengerIDA;
    public int passengerIDB;
    public int numberOfFlights;

    public PassengerPair(int passengerIDA, int passengerIDB, 
    int numberOfFlights) {

        this.passengerIDA = passengerIDA;
        this.passengerIDB = passengerIDB;
        this.numberOfFlights = numberOfFlights;
    }

    public int compareTo(PassengerPair other) {

        // Sorts passenger pairs by number of flights shared ascending 
        return (this.numberOfFlights - other.numberOfFlights);
    }
}

public class FlightData {

    private static String passengerFilename = "passengers.csv";
    private static String flightDataFilename = "flightData.csv";
    private static String question1Filename = "javaQuestion1.csv";
    private static String question2Filename = "javaQuestion2.csv";
    private static String question3Filename = "javaQuestion3.csv";
    private static String question4Filename = "javaQuestion4.csv";
    private static String question5Filename = "javaQuestion5.csv";

    public static void main(String[] args) {
        
        Map<Integer, Passenger> passengers = getPassengers(passengerFilename);
        Map<Integer, Flight> flights = getFlights(flightDataFilename);

        getPassengersOfEachFlight(passengers, flights, flightDataFilename);

        getTotalFlightsForEachMonth(passengers, flights, question1Filename);
        getMostFrequentFlyers(passengers, flights, 100, question2Filename);
        getPassengerRuns(passengers, flights, "uk", question3Filename);
        getPassengersWithSharedFlights(passengers, flights, 3, question4Filename);
    }

    public static Map<Integer, Passenger> getPassengers(String filename) {

        // Returns a map of (passengerID, passenger) pairs 
        // from the given input filename 

        Map<Integer, Passenger> passengers = new HashMap<>();

        try {
            
            Scanner scanner = new Scanner(new File(filename));

            // Read header 
            scanner.nextLine();

            while (scanner.hasNextLine()) {

                // Read passenger line 
                String[] passenger = scanner.nextLine().split(",");

                int passengerID = Integer.parseInt(passenger[0]);
                String firstName = passenger[1];
                String lastName = passenger[2];
                
                passengers.put(passengerID, 
                new Passenger(passengerID, firstName, lastName));
            }
        } catch (FileNotFoundException ex) {

            System.out.println(ex.getMessage());
        }

        return passengers;
    }

    public static Map<Integer, Flight> getFlights(String filename) {

        // Returns a map of (flightID, flight) pairs 
        // from the given input filename 

        Map<Integer, Flight> flights = new HashMap<>();

        try {
            
            Scanner scanner = new Scanner(new File(filename));

            // Read header 
            scanner.nextLine();

            while (scanner.hasNextLine()) {

                // Read flight line 
                String[] flight = scanner.nextLine().split(",");

                int passengerID = Integer.parseInt(flight[0]);
                int flightID = Integer.parseInt(flight[1]);
                String from = flight[2];
                String to = flight[3];
                String[] date = flight[4].split("-");
                int year = Integer.parseInt(date[0]);
                int month = Integer.parseInt(date[1]);
                int day = Integer.parseInt(date[2]);
                
                flights.put(flightID, 
                new Flight(flightID, from, to, 
                new Date(year, month, day)));
            }

            scanner.close();
        } catch (FileNotFoundException ex) {

            System.out.println(ex.getMessage());
        }

        return flights;
    }

    public static void getPassengersOfEachFlight(
    Map<Integer, Passenger> passengers, 
    Map<Integer, Flight> flights, String filename) {

        // Creates a set of passengers that travelled on 
        // each flight, as well as a set of flights that 
        // each passenger travelled on 

        try {
        
            Scanner scanner = new Scanner(new File(filename));

            // Read header 
            scanner.nextLine();

            while (scanner.hasNextLine()) {

                // Read flight line 
                String[] flight = scanner.nextLine().split(",");

                int passengerID = Integer.parseInt(flight[0]);
                int flightID = Integer.parseInt(flight[1]);
                
                // Update the relevant passenger 
                passengers.get(passengerID).flights
                .add(flights.get(flightID));

                // Update the relevant flight 
                flights.get(flightID).passengers
                .add(passengers.get(passengerID));
            }

            scanner.close();
        } catch (FileNotFoundException ex) {

            System.out.println(ex.getMessage());
        }
    }

    public static void getTotalFlightsForEachMonth(
    Map<Integer, Passenger> passengers, 
    Map<Integer, Flight> flights, String filename) {

        // Outputs the total flights for each month 
        // to the given output filename 

        int[] monthNumberOfFlights = new int[12];

        for (Integer flightID : flights.keySet()) {

            int month = flights.get(flightID).date.month;

            monthNumberOfFlights[month - 1]++;
        }

        try {
            PrintWriter writer = new PrintWriter(new File(filename));

            String header = "month,numberOfFlights";
            writer.println(header);

            for (int i = 0; i < monthNumberOfFlights.length; i++) {

                // Write line 
                int month = i + 1;
                String line = month + "," + monthNumberOfFlights[i];
                writer.println(line);
            }

            writer.close();
        } catch (FileNotFoundException ex) {

            System.out.println(ex.getMessage());
        }
    }

    public static void getMostFrequentFlyers(
    Map<Integer, Passenger> passengers, 
    Map<Integer, Flight> flights, int k, String filename) {

        // Outputs the k most frequent flyers 
        // to the given output filename 

        // Create min heap of at most k elements 
        PriorityQueue<Passenger> bestPassengers = new PriorityQueue<>();

        for (Passenger passenger : passengers.values()) {

            bestPassengers.add(passenger);

            // Ensure min heap has at most k elements 
            if (bestPassengers.size() > k) {

                // Remove the (k + 1)th most frequent passenger 
                bestPassengers.poll();
            }
        }

        // Extract the passengers in ascending order 
        List<Passenger> outputPassengers = new ArrayList<>();

        while (!bestPassengers.isEmpty()) {

            outputPassengers.add(bestPassengers.poll());
        }

        Collections.reverse(outputPassengers);

        // Output the top k passengers in order of 
        // number of flights descending 
        try {
            PrintWriter writer = new PrintWriter(new File(filename));

            String header = "passengerId,numberOfFlights," 
            + "firstName,lastName";
            writer.println(header);

            for (Passenger passenger : outputPassengers) {

                // Write line 
                String line = passenger.passengerID + ","
                + passenger.flights.size() + ","
                + passenger.firstName + ","
                + passenger.lastName;
                writer.println(line);
            }

            writer.close();
        } catch (FileNotFoundException ex) {

            System.out.println(ex.getMessage());
        }
    }

    public static void getPassengerRuns(
    Map<Integer, Passenger> passengers, 
    Map<Integer, Flight> flights, String runEnd, String filename) {

        // Outputs the longest runs for each passenger 
        // given a location runEnd 
        // to the given output filename 

        List<PassengerRun> passengerRuns = new ArrayList<>();

        for (int passengerID : passengers.keySet()) {

            // Create a list of flights sortable by date 
            List<Flight> passengerFlights = 
            new ArrayList<>(passengers.get(passengerID).flights);

            int longestRun = getLongestRun(passengerFlights, runEnd);

            passengerRuns.add(new PassengerRun(passengerID, longestRun));
        }

        // Sort passenger runs by longest run descending 
        Collections.sort(passengerRuns);
        Collections.reverse(passengerRuns);

        // Output passenger runs 
        try {
            PrintWriter writer = new PrintWriter(new File(filename));

            String header = "passengerId,longestRun";
            writer.println(header);

            for (PassengerRun passengerRun : passengerRuns) {

                // Write line 
                String line = passengerRun.passengerID + "," 
                + passengerRun.longestRun;
                writer.println(line);
            }

            writer.close();
        } catch (FileNotFoundException ex) {

            System.out.println(ex.getMessage());
        }
    }

    public static int getLongestRun(List<Flight> passengerFlights, 
    String runEnd) {

        // Returns the longest run 
        // given a list of passenger flights 
        // and a location runEnd 

        // Sort passenger flights by date ascending 
        Collections.sort(passengerFlights);

        int longestRun = 0;
        int currentRun = 0;

        if (!passengerFlights.get(0).from.equals(runEnd)) {

            // Start at first location 
            longestRun++;
            currentRun++;
        }

        for (Flight flight : passengerFlights) {

            if (flight.to.equals(runEnd)) {

                // End of current run 
                currentRun = 0;

                continue;
            }

            // Extend current run 
            currentRun++;
            longestRun = Math.max(longestRun, currentRun);
        }

        return longestRun;
    }

    public static void getPassengersWithSharedFlights(
    Map<Integer, Passenger> passengers, 
    Map<Integer, Flight> flights, int k, String filename) {

        // Outputs the passenger pairs sharing > k flights 
        // to the given output filename 

        // passengerPairIDNumberOfFlights is a set of 
        // (passengerPairID, numberOfFlights) pairs 
        Map<String, Integer> passengerPairIDNumberOfFlights = new HashMap<>();

        for (int passengerIDA : passengers.keySet()) {

            for (int passengerIDB : passengers.keySet()) {

                if (passengerIDA >= passengerIDB) {

                    // Same passenger is both parts of the pair 
                    // Ensures passenger A has a lower ID than 
                    // passenger B to avoid double counting 
                    continue;
                }

                // Create the ID representing the pair 
                String passengerPairID = passengerIDA + "&" + passengerIDB;

                // Get number of shared flights 
                int numberOfFlights = 0;

                for (Flight flightA : passengers.get(passengerIDA)
                .flights) {

                    if (passengers.get(passengerIDB).flights
                    .contains(flightA)) {

                        // Current flight is shared 
                        numberOfFlights++;
                    }
                }

                passengerPairIDNumberOfFlights.put(passengerPairID, 
                numberOfFlights);
            }
        }

        List<PassengerPair> passengerPairs = new ArrayList<>();

        for (String passengerPairID : passengerPairIDNumberOfFlights.keySet()) {

            // Create a list of passenger pairs sortable 
            // by number of flights shared 
            String[] passengerIDs = passengerPairID.split("&");
            int passengerIDA = Integer.parseInt(passengerIDs[0]);
            int passengerIDB = Integer.parseInt(passengerIDs[1]);
            int numberOfFlights = passengerPairIDNumberOfFlights
            .get(passengerPairID);

            passengerPairs.add(new PassengerPair(passengerIDA, 
            passengerIDB, numberOfFlights));
        }

        Collections.sort(passengerPairs);
        Collections.reverse(passengerPairs);

        // Output passenger pairs 
        try {
            PrintWriter writer = new PrintWriter(new File(filename));

            String header = "passenger1Id,passenger2Id," 
            + "numberOfFlightsTogether";
            writer.println(header);

            for (PassengerPair passengerPair : passengerPairs) {

                if (passengerPair.numberOfFlights <= k) {

                    // Number of flights shared must be more than k 
                    continue;
                }

                // Write line 
                String line = passengerPair.passengerIDA + "," 
                + passengerPair.passengerIDB + "," 
                + passengerPair.numberOfFlights;
                writer.println(line);
            }

            writer.close();
        } catch (FileNotFoundException ex) {

            System.out.println(ex.getMessage());
        }
    }
}
